### ISSUE TITLE
Request Block

### ISSUE THREAD

**Issue by abhij89007** [mebjas/CSRF-Protector-PHP#67]
Hi,

I have a form:
<form id="register_form_new">
	<h4>Register</h4>
		<div id="registerFromLogin">
			<div class="col-md-10 col-md-offset-1 no-margin text-left">
				<label class="pull-left">Full Name*</label><br>
				<input type="text" name="full_name_new" class="form-control margin-bottom-15" id="full_name_new" required>
				<p id="errorname" style="display:none;">Please enter your name</p>
										
                                <label class="pull-left">Email Address*</label><br>
				<input type="text"name="email_new" id="email_new" class="form-control margin-bottom-15" required>
				<p id="erroremail" style="display:none;">Please enter your email</p>

                               <button type="button" onclick="validateform();" class="btn btn-secondary btn-md auto-width">Sign up</button>
			</div>

		</div>
</form>

What this form do is try to validate inputs before submission. And then initiate an ajax call to submit form input, but this module code seems to be blocking it. It blocks the original request, initiates its own, and ends up in both requests being blocked. When i disable this module, it works just fine. Attached a screenshot.

Please suggest a solution.

![csrfprotector](https://cloud.githubusercontent.com/assets/15982920/23811543/811e8c34-05fc-11e7-897b-2a52fa369af3.png)

**Comment by mebjas** [comment]
@abhij89007 - could you provide more details? Browser, OS, maybe more steps to reproduce.

Also have you included the library to both calling and receiving page of AJAX request?

**Comment by abhij89007** [comment]
@mebjas Browser is Firefox, OS is windows 7. I have installed library using composer.
Yes i have included the library at both pages. It works fine in case of POST method but not in case of GET method.
Steps for how i got this error:
1. Add library on signup page(calling page) and submission page(receiving page).
2. Submit the inputs on the form.
3. Add a validate function on submit button of the form.
4. Validate the input of fields, and if fine use AJAX to create a GET request to send form data to receiving page.
5. And you will see the error for which i have attached a screenshot in my previous comment.

**Comment by abhij89007** [comment]
Also i trying this in private window of Firefox to avoid any cache, if this helps.

**Comment by mebjas** [comment]
@abhij89007 have you configured the verify get for config param for the request url? https://github.com/mebjas/CSRF-Protector-PHP/blob/master/libs/config.sample.php#L26

More information - https://github.com/mebjas/CSRF-Protector-PHP/wiki/Configurations

**Comment by abhij89007** [comment]
I have left it as blank array, is it mandatory to have something in it? Does it mandatorily have values, documentation didn't say so.

**Comment by mebjas** [comment]
Yes, GET requests are not supported by default. **It cannot be supported by default**. If there were default support for GET request, you'd not be able to even load landing page, as it has no token to send.

If you add the url of the ajax to this array, or a pattern there - it should work.

**Comment by abhij89007** [comment]
Ok i will give this a try. But shouldn't it work for all ajax requests by default? be it POST or GET.

**Comment by mebjas** [comment]
No Get requests are ideally used for retrieving data. That's not where CSRF is applicable.

**Comment by abhij89007** [comment]
Ok, so why is that AJAX request using GET method is not working. It should work though, even if it doesn't add the CSRF token in it, right? Since i haven't added any rules or regex in verify GET array. It should let it work without adding any security token or blocking it right away.

**Comment by mebjas** [comment]
Yeah, that is true. Ideally this should not be blocked.
I'll see if I can reproduce this bug.

----------------------------------------------------------------------------
Kind Regards,
Minhaz
minhazav.xyz | Blog <http://blog.minhazav.xyz> | Projects
<http://github.com/mebjas> | LinkedIn <https://in.linkedin.com/in/minhazav>

**Comment by abhij89007** [comment]
I am maintaining a legacy project, and it is using hundreds of AJAX GET requests to retrieve or send data. Now adding this library will block all those requests. It should just skip those requests if not specified in verify get array. Changing all of them to POST will be a huge task. And not just me , many user may face this situation. Please update this. Thanks.

**Comment by abhij89007** [comment]
@mebjas I have defined the rule for verify get in config file. 
`	 "verifyGetFor" => array('*://*/exe/*')`

I tried it with same scenario as described in one of my previous comment.

1. Add library on signup page(calling page) and submission page(receiving page).
2. Submit the inputs on the form.
3. Add a validate function on submit button of the form.
4. Validate the input of fields, and if fine use AJAX to create a GET request to send form data to receiving page.
It isn't adding any token in the get request and ending up in 403 error, but surprisingly when i resend the same request using console(firebug), it adds the token in the request and works just fine. How come is it behaving like this??

![2017-03-15_2039](https://cloud.githubusercontent.com/assets/15982920/23955382/b826ea20-09bf-11e7-92d6-148494d854dc.png)

**Comment by mebjas** [comment]
@abhij89007 you are using preventdefault() to stop form from submitting itself?

**Comment by abhij89007** [comment]
@mebjas I apologies, i mentioned it wrong earlier, i am not using a submit button, its just a button. As you can check my form i added in my first description.

**Comment by abhij89007** [comment]
@mebjas I have found another bug, library seems not to be working on relative paths. So if url is giving like this:

`url: "/exe/new-user-registration.php",` it works, but if its defined like this `url: "../exe/new-user-registration.php",` it ends up in error.

**Comment by mebjas** [comment]
@abhij89007 - I tested both the bugs you mentioned and I'm not able to reproduce either of them. Both work for me. Also, `verifyGetFor` flag is used to enable CSRF mitigation for get requests, it should be 403 free when it's empty.

**Comment by abhij89007** [comment]
Ok figured why is it happening. It is the path issue. 
Like i mentioned in previous comment, if url is defined like this:
`url: "/exe/new-user-registration.php"`, it works,
but if its defined like this `url: "../exe/new-user-registration.php"`, it ends up in error. This is with both POST and GET method.

**Comment by abhij89007** [comment]
also is it possible to have the cookieexpirytime in config file instead of going to edit the vendor library, reason being updating composer , updates any changes made in the library..

**Comment by mebjas** [comment]
yeah there has been an ask like that. There were some PR on that, would try to get that up and merge.

Can I consider this as not-bug and set as resolved?

**Comment by mebjas** [comment]
@abhij89007 ?

**Comment by abhij89007** [comment]
@mebjas thank you for considering adding cookieexpirytime in config file . But like i mentioned in my earlier comments, the path issue still exists.
I have defined the rule for verify get in config file.
"verifyGetFor" => array('*://*/exe/*')

if url is defined like this:
`url: "/exe/new-user-registration.php"`, it works,
but if its defined like this `url: "../exe/new-user-registration.php"`, it ends up in error. This is with both POST and GET method.

**Comment by abhij89007** [comment]
triggering click using jquery doesn't append csrf token in the ajax request bind with that triggered click. Am i missing something here?

**Comment by abhij89007** [comment]
Also every now and then user complains about 403 forbidden error on the site, is this because token expired or can it be something else?

**Comment by mebjas** [comment]
@abhij89 does this issue still exist?

**Comment by abhij89** [comment]
Yes, sorry for delayed response. It's still the same as explained above.
